### PR TITLE
[#23] Create endpoint /api/rest/report/tag

### LINKED ISSUE #23: Create /api/rest/report/tag endpoint

### PR DESCRIPTION
**Pull request by marsofandrew**
close #23

### CODE REVIEW

File: api/src/main/java/org/confeti/controllers/core/ReportController.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+@RequiredArgsConstructor
+public class ReportController {
+
+    private static final String CONFERENCE_NAME_URI_PARAMETER = "conference_name";
+    private static final String YEAR_URI_PARAMETER = "year";
+    private final ReportService reportService;
+
+    @GetMapping(path = "/tag", params = {YEAR_URI_PARAMETER, CONFERENCE_NAME_URI_PARAMETER})
+    @ResponseBody
+    public Mono<ResponseEntity<?>> handleTagRequest(@RequestParam(YEAR_URI_PARAMETER) final int year,
```
Comment (lapter57) [inline api/src/main/java/org/confeti/controllers/core/ReportController.java:38]: You can create a method to count the tags of conference reports of a specific year and reuse it in other methods.
```
    private static Mono<Map<String, Long>> countTags(final Flux<Report> reports) {
        return reports.flatMapIterable(report -> Objects.requireNonNullElse(report.getTags(), Collections.emptySet()))
                .groupBy(s -> s)
                .flatMap(group -> Mono.zip(Mono.just(Objects.requireNonNull(group.key())), group.count()))
                .collectMap(Tuple2::getT1, Tuple2::getT2);
    }

    @GetMapping(path = "/tag", params = {YEAR_URI_PARAMETER, CONFERENCE_NAME_URI_PARAMETER})
    @ResponseBody
    public Mono<ResponseEntity<?>> handleTagRequest(@RequestParam(YEAR_URI_PARAMETER) final int year,
                                                    @RequestParam(CONFERENCE_NAME_URI_PARAMETER) final String conferenceName) {
        return countTags(reportService.findBy(conferenceName, year))
                .<ResponseEntity<?>>map(map -> ResponseEntity.ok(new TagResponse(conferenceName, Map.of(year, map))))
                .onErrorResume(Exception.class, err -> Mono.just(ResponseEntity.badRequest().body(new ErrorResponse(err.getMessage()))));
    }
```

File: api/src/main/java/org/confeti/controllers/core/ReportController.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                .map(report -> Objects.requireNonNullElse(report.getTags(), Collections.<String>emptySet()))
+                .collect(Suppliers.ofInstance(new ArrayList<String>()), ArrayList::addAll)
+                .map(ReportController::convertCollectionToMap)
+                .<ResponseEntity<?>>map(map -> ResponseEntity.ok(new TagResponse(conferenceName, Map.of(year, map))))
+                .onErrorResume(Exception.class, err -> Mono.just(ResponseEntity.badRequest().body(new ErrorResponse(err.getMessage()))));
+    }
+
+    @GetMapping(path = "/tag", params = {CONFERENCE_NAME_URI_PARAMETER})
+    @ResponseBody
+    public Mono<ResponseEntity<?>> handleTagRequest(@RequestParam(CONFERENCE_NAME_URI_PARAMETER) final String conferenceName) {
```
Comment (lapter57) [inline api/src/main/java/org/confeti/controllers/core/ReportController.java:47]: You can create a method to count the tags of conference reports for all years using `countTags`:
```
    private static Mono<Map<Integer, Map<String, Long>>> countTagsByYear(final Flux<GroupedFlux<Integer, Report>> reports) {
        return reports.flatMap(group -> Mono.zip(Mono.just(Objects.requireNonNull(group.key())), countTags(group)))
                .collectMap(Tuple2::getT1, Tuple2::getT2);
    }

    @GetMapping(path = "/tag", params = {CONFERENCE_NAME_URI_PARAMETER})
    @ResponseBody
    public Mono<ResponseEntity<?>> handleTagRequest(@RequestParam(CONFERENCE_NAME_URI_PARAMETER) final String conferenceName) {
        return countTagsByYear(reportService.findBy(conferenceName)
                    .groupBy(report -> report.getConferences().iterator().next().getYear()))
                .<ResponseEntity<?>>map(map -> ResponseEntity.ok(new TagResponse(conferenceName, map)))
                .onErrorResume(Exception.class, err -> Mono.just(ResponseEntity.badRequest().body(new ErrorResponse(err.getMessage()))));
    }
```

File: api/src/main/java/org/confeti/controllers/core/ReportController.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+                                .forEach(year -> createYearMap(collection, () -> year, report::getTags))
+                )
+                .map(map -> new TagResponse(conferenceName, map))
+                .<ResponseEntity<?>>map(ResponseEntity::ok)
+                .onErrorResume(Exception.class, err -> Mono.just(ResponseEntity.badRequest().body(new ErrorResponse(err.getMessage()))));
+    }
+
+    @GetMapping(path = "/tag")
+    @ResponseBody
+    public Mono<ResponseEntity<?>> handleTagRequest(
```
Comment (lapter57) [inline api/src/main/java/org/confeti/controllers/core/ReportController.java:57]: You can simplify this code just use `groupBy` and reuse the method `countTagsByYear`:
```
    @GetMapping(path = "/tag")
    @ResponseBody
    public Mono<ResponseEntity<?>> handleTagRequest(
            @RequestParam(value = YEAR_URI_PARAMETER, required = false) final Optional<Integer> year) {
        return reportService.findAll()
                .flatMap(report -> Flux.fromIterable(report.getConferences())
                        .filter(conference -> year.isEmpty() || year.get().equals(conference.getYear()))
                        .zipWith(Mono.just(report)))
                .groupBy(t -> t.getT1().getName())
                .flatMap(group -> countTagsByYear(group.groupBy(t -> t.getT1().getYear(), Tuple2::getT2))
                        .map(map -> new TagResponse(group.key(), map)))
                .collectList()
                .<ResponseEntity<?>>map(ResponseEntity::ok)
                .onErrorResume(Exception.class, err -> Mono.just(ResponseEntity.badRequest().body(new ErrorResponse(err.getMessage()))));
    }
```

File: api/src/main/java/org/confeti/controllers/dto/core/TagResponse.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+import java.util.Map;
+
+@Accessors(chain = true)
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class TagResponse {
+    private String conferenceName;
+    private Map<Integer, Map<String, Integer>> years;
```
Comment (lapter57) [inline api/src/main/java/org/confeti/controllers/dto/core/TagResponse.java]: The number of tags must be of type `Long`:  `Map<Integer, Map<String, Long>>`

File: api/src/test/java/org/confeti/controllers/core/ReportControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+                                .tags(Set.of("a", "b"))
+                                .conferences(Set.of(conference1))
+                                .build(),
+                        Report.builder("title11")
+                                .tags(Set.of("a", "c"))
+                                .conferences(Set.of(conference1))
+                                .build(),
+                        Report.builder("title2")
+                                .tags(Set.of("a", "c"))
+                                .conferences(Set.of(conference2, Conference.builder("tttt", 1975).build()))
```
Comment (lapter57) [inline api/src/test/java/org/confeti/controllers/core/ReportControllerTest.java:81]: There is no `conferences` field in the [report_by_conference](https://github.com/confeti/confeti/blob/develop/api/src/main/resources/cassandra/ddl-schema.cql#L127) table, so when you search for reports by conference name, the set of conferences of report will have only one conference.

Comment (marsofandrew) [reply api/src/test/java/org/confeti/controllers/core/ReportControllerTest.java:81]: nethertheless, we could check such scenario.

File: api/src/main/java/org/confeti/controllers/core/ReportController.java (hunk 1 above)
Comment (marsofandrew) [reply api/src/main/java/org/confeti/controllers/core/ReportController.java:38]: Agree

**Review by lapter57 — APPROVED** [review]

**Pull request merged**